### PR TITLE
fix: Stub instance of the same component

### LINKED ISSUE #1162: shallowMount doesn't stub instances of the same component
# Expected behavior

When using `shallowMount` on a component that contains itself, nested instances should be stubbed, so the output for the reproduction in the shared repo should be something like:

```
    <ul>
      <li>Option 1
        <!--v-if-->
      </li>
      <li>Option 2 <list-stub></list-stub>
      </li>
      <li>Option 3
        <!--v-if-->
      </li>
    </ul>
```

# Current behavior

Nested instances are being mounted as if we were using `mount`. Current output is:

```
    <ul>
      <li>Option 1
        <!--v-if-->
      </li>
      <li>Option 2 <ul>
          <li>Option 2.1
            <!--v-if-->
          </li>
          <li>Option 2.2
            <!--v-if-->
          </li>
        </ul>
      </li>
      <li>Option 3
        <!--v-if-->
      </li>
    </ul>
```

# Reproduction:

https://github.com/victorborrasdev/vue-test-utils-nested-stubs

Please let me know if you need any additional info. Thanks!

### PR DESCRIPTION
**Pull request by freakzlike**
Fixes #1162 

I'm not really happy with the wording, because there is a lot of double negation. However, the code does the job

### CODE REVIEW

File: src/vnodeTransformers/stubComponentsTransformer.ts [hunk 1]
```diff
@@ -36,10 +36,14 @@ interface StubOptions {
...
+const doNotStubComponents = new WeakMap<ConcreteComponent, boolean>()
 const shouldNotStub = (type: ConcreteComponent) => doNotStubComponents.has(type)
-export const addToDoNotStubComponents = (type: ConcreteComponent) =>
-  doNotStubComponents.add(type)
+const shouldNotStubRoot = (type: ConcreteComponent) =>
+  !!doNotStubComponents.get(type)
+export const addToDoNotStubComponents = (
+  type: ConcreteComponent,
+  onlyRoot?: boolean
+) => doNotStubComponents.set(type, !!onlyRoot)
```
Comment (cexbrayat) [inline src/vnodeTransformers/stubComponentsTransformer.ts:46]: maybe we could have two distinct functions, with different names to make it easier to understand?
like `addToDoNotStubComponentsButOnlyInRoot`? (I admit it's not a great name ^^)

**Review by cexbrayat — APPROVED** [review]
Thanks for taking a stab at this @freakzlike 
Code looks good, even if I agree it's hard to understand all the negations of negations

**Comment by xanf** [comment]
I'm iterating on this one, and while I'm working on suggestion, would like to share my thoughts:

* we are already tracking "root" instances for auto-destroy purposes, so what I'm thinking might be is drop `addToDoNotStubComponents` logic for root but instead when making stub logic check it from registry of "root" instances (and move `trackInstance` logic and reigstry to separate f ile

**Comment by freakzlike** [comment]
> I'm iterating on this one, and while I'm working on suggestion, would like to share my thoughts:
> 
> * we are already tracking "root" instances for auto-destroy purposes, so what I'm thinking might be is drop `addToDoNotStubComponents` logic for root but instead when making stub logic check it from registry of "root" instances (and move `trackInstance` logic and reigstry to separate f ile

You are right. It's only used for the root components and it might be a better way

**Comment by freakzlike** [comment]
I tried to retrieve or check for the root component without the need of `rootComponents`, but didn't manage to do so for all cases. The logic has not changed now, but it's a bit easier to read I think

**Comment by lmiller1990** [comment]
Need a rebase, since we just released the `renderToString` PR.

Glad @xanf is running this on a real code base - will wait back to hear from him before merging this one up.

File: src/vnodeTransformers/stubComponentsTransformer.ts [hunk 2]
```diff
@@ -102,13 +97,18 @@ const resolveComponentStubByName = (
   }
 }
 
-interface CreateStubComponentsTransformerConfig {
+export interface CreateStubComponentsTransformerConfig {
+  rootComponents: {
+    legacy?: Component
+    component?: Component
+  }
```
Comment (cexbrayat) [inline src/vnodeTransformers/stubComponentsTransformer.ts:104]: Would you mind adding a few comments on what each field does?

**Review by cexbrayat — APPROVED** [review]
LGTM

This needs to be rebased, as some logic moved to `src/createInstance` now that `renderToString` landed (sorry)

**Comment by lmiller1990** [comment]
Shall we merge or did we want more testing from @xanf? Happy to move forward as-is.

**Comment by xanf** [comment]
@freakzlike @lmiller1990 I'm going to finish testing no later than tomorrow, sorry for delay - our codebase uses v-once, and I need to use vuejs with this PR https://github.com/vuejs/core/pull/7730 merged to test it, cause compiler blows up

**Comment by lmiller1990** [comment]
I think we should merge and release this one. I'll give it another 24h then ship.

**Pull request merged**